File: prototype/src/wet_run/combat/run_mutators.py
"""Run Mutators system (ADR-0163).

5+ optional mutators applied at run start that change the rules
per-playthrough. Mutators amplify Pillar 3 (death weight) and
Pillar 4 (build variety).
"""

from __future__ import annotations

from enum import StrEnum
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ..engine.state import AppState


class RunMutator(StrEnum):
    """Run modifiers that change game rules per-playthrough."""

    LOW_HP = "low_hp"
    DOUBLE_ALARM = "double_alarm"
    ICE_X2 = "ice_x2"
    NO_HEAL = "no_heal"
    STEALTH_ONLY = "stealth_only"


MUTATORS: dict[RunMutator, dict[str, str]] = {
    RunMutator.LOW_HP: {
        "name": "FRAGILE WETWARE",
        "description": "Start with 50% max HP. One bad run ends you.",
        "icon": "low_hp",
    },
    RunMutator.DOUBLE_ALARM: {
        "name": "HOT TRACE",
        "description": "Alarm ticks 2x faster. ICE pursues.",
        "icon": "double_alarm",
    },
    RunMutator.ICE_X2: {
        "name": "POPULATED GRID",
        "description": "Every encounter is 1v2 or 1v3. The grid is *crowded*.",
        "icon": "ice_x2",
    },
    RunMutator.NO_HEAL: {
        "name": "DEAD MAN WALKING",
        "description": "Cannot salvage HEAL from kills. No recovery.",
        "icon": "no_heal",
    },
    RunMutator.STEALTH_ONLY: {
        "name": "GHOST PROTOCOL",
        "description": "Only stealth skills available. Silent runs.",
        "icon": "stealth_only",
    },
}


def get_mutator_info(mutator: RunMutator) -> dict[str, str]:
    """Return display info for a mutator."""
    return MUTATORS[mutator]


def apply_mutators(app_state: AppState, mutators: list[RunMutator]) -> None:
    """Apply mutators to AppState (idempotent — clears previous first)."""
    clear_mutators(app_state)
    for mutator in mutators:
        if mutator == RunMutator.LOW_HP:
            app_state.player_max_hp = app_state.player_max_hp // 2
            app_state.player_hp = min(app_state.player_hp, app_state.player_max_hp)
        elif mutator == RunMutator.DOUBLE_ALARM:
            app_state.alarm_speed_multiplier = 2.0
        elif mutator == RunMutator.ICE_X2:
            app_state.encounter_multiplier = 2
        elif mutator == RunMutator.NO_HEAL:
            app_state.heal_disabled = True
        elif mutator == RunMutator.STEALTH_ONLY:
            app_state.skill_filter = "stealth_only"
    app_state.active_mutators = tuple(mutators)


def clear_mutators(app_state: AppState) -> None:
    """Clear all mutator effects from AppState."""
    for mutator in app_state.active_mutators:
        if mutator == RunMutator.LOW_HP:
            app_state.player_max_hp = app_state.player_max_hp * 2
        elif mutator == RunMutator.DOUBLE_ALARM:
            app_state.alarm_speed_multiplier = 1.0
        elif mutator == RunMutator.ICE_X2:
            app_state.encounter_multiplier = 1
        elif mutator == RunMutator.NO_HEAL:
            app_state.heal_disabled = False
        elif mutator == RunMutator.STEALTH_ONLY:
            app_state.skill_filter = None
    app_state.active_mutators = ()


def is_mutator_active(app_state: AppState, mutator: RunMutator) -> bool:
    """Check if a mutator is currently active."""
    return mutator in app_state.active_mutators


def get_active_mutators(app_state: AppState) -> tuple[RunMutator, ...]:
    """Return the active mutators list."""
    result: tuple[RunMutator, ...] = tuple(RunMutator(m) for m in app_state.active_mutators)
    return result


def get_alarm_multiplier(app_state: AppState) -> float:
    """Return the alarm speed multiplier (1.0 = normal, 2.0 = DOUBLE_ALARM)."""
    return app_state.alarm_speed_multiplier


def get_encounter_multiplier(app_state: AppState) -> int:
    """Return the encounter count multiplier (1 = normal, 2 = ICE_X2)."""
    return app_state.encounter_multiplier


def is_heal_disabled(app_state: object) -> bool:
    """Return True if HEAL salvage is disabled (NO_HEAL mutator)."""
    return bool(getattr(app_state, "heal_disabled", False))


def is_stealth_only(app_state: AppState) -> bool:
    """Return True if only stealth skills available (STEALTH_ONLY mutator)."""
    return app_state.skill_filter == "stealth_only"


def hp_multiplier(mutator: RunMutator) -> float:
    """Return HP multiplier for a mutator (LOW_HP = 0.5, others = 1.0)."""
    if mutator == RunMutator.LOW_HP:
        return 0.5
    return 1.0


__all__ = [
    "MUTATORS",
    "RunMutator",
    "apply_mutators",
    "clear_mutators",
    "get_active_mutators",
    "get_alarm_multiplier",
    "get_encounter_multiplier",
    "get_mutator_info",
    "hp_multiplier",
    "is_heal_disabled",
    "is_mutator_active",
    "is_stealth_only",
]

File: web/src/core/run_mutators.ts
/** Run Mutators system (ADR-0163 port).
 *
 * 5 optional mutators applied at run start that change the rules per
 * playthrough. Mutators amplify Pillar 3 (death weight) and Pillar 4
 * (build variety).
 *
 *   - LOW_HP        — Start with 50% max HP. One bad run ends you.
 *   - DOUBLE_ALARM  — Alarm ticks 2x faster. ICE pursues.
 *   - ICE_X2        — Every encounter is 1v2 or 1v3. The grid is crowded.
 *   - NO_HEAL       — Cannot salvage HEAL from kills. No recovery.
 *   - STEALTH_ONLY  — Only stealth skills available. Silent runs.
 *
 * The mutator system operates on a {@link MutableRunState} target — a
 * mutable bag of fields the mutators can read and write. This mirrors
 * the Python ``apply_mutators(app_state, mutators)`` signature without
 * coupling to a specific app-state shape.
 */

/** All available mutator kinds. */
export type RunMutator =
  | "low_hp"
  | "double_alarm"
  | "ice_x2"
  | "no_heal"
  | "stealth_only";

/** Display info for a single mutator. */
export interface MutatorInfo {
  readonly id: RunMutator;
  readonly name: string;
  readonly description: string;
  readonly icon: string;
}

/**
 * Mutable run-state fields that mutators read/write.
 *
 * Mirrors the Python `AppState` fields touched by `apply_mutators` /
 * `clear_mutators`. The application constructs one of these from its
 * full state and passes it in; mutators never need to know about other
 * fields.
 */
export interface MutableRunState {
  /** Current player HP. Mutators clamp to `playerMaxHp` after a change. */
  playerHp: number;
  /** Player max HP. LOW_HP halves this on apply, doubles on clear. */
  playerMaxHp: number;
  /** Alarm tick multiplier (1.0 = normal, 2.0 = DOUBLE_ALARM). */
  alarmSpeedMultiplier: number;
  /** Per-encounter enemy count multiplier (1 = normal, 2 = ICE_X2). */
  encounterMultiplier: number;
  /** True when HEAL salvage is disabled (NO_HEAL mutator). */
  healDisabled: boolean;
  /** Active skill filter; null = no filter, "stealth_only" = STEALTH_ONLY. */
  skillFilter: string | null;
  /** The currently active mutators (frozen tuple equivalent). */
  activeMutators: ReadonlyArray<RunMutator>;
}

/** Full catalog of mutators and their display info. */
export const MUTATORS: Readonly<Record<RunMutator, MutatorInfo>> = Object.freeze({
  low_hp: Object.freeze({
    id: "low_hp",
    name: "FRAGILE WETWARE",
    description: "Start with 50% max HP. One bad run ends you.",
    icon: "low_hp",
  }),
  double_alarm: Object.freeze({
    id: "double_alarm",
    name: "HOT TRACE",
    description: "Alarm ticks 2x faster. ICE pursues.",
    icon: "double_alarm",
  }),
  ice_x2: Object.freeze({
    id: "ice_x2",
    name: "POPULATED GRID",
    description: "Every encounter is 1v2 or 1v3. The grid is *crowded*.",
    icon: "ice_x2",
  }),
  no_heal: Object.freeze({
    id: "no_heal",
    name: "DEAD MAN WALKING",
    description: "Cannot salvage HEAL from kills. No recovery.",
    icon: "no_heal",
  }),
  stealth_only: Object.freeze({
    id: "stealth_only",
    name: "GHOST PROTOCOL",
    description: "Only stealth skills available. Silent runs.",
    icon: "stealth_only",
  }),
});

/** All mutator IDs in declaration order. */
export const ALL_MUTATORS: ReadonlyArray<RunMutator> = Object.freeze([
  "low_hp",
  "double_alarm",
  "ice_x2",
  "no_heal",
  "stealth_only",
]);

/** Return display info for a mutator (throws if the ID is unknown). */
export function getMutatorInfo(mutator: RunMutator): MutatorInfo {
  return MUTATORS[mutator];
}

/** HP multiplier for a mutator (LOW_HP = 0.5, others = 1.0). */
export function hpMultiplier(mutator: RunMutator): number {
  return mutator === "low_hp" ? 0.5 : 1.0;
}

/** Apply mutators to a run state (idempotent — clears previous first). */
export function applyMutators(
  state: MutableRunState,
  mutators: ReadonlyArray<RunMutator>,
): void {
  clearMutators(state);
  for (const mutator of mutators) {
    switch (mutator) {
      case "low_hp":
        // Guard against playerMaxHp = 0 default: if 0, initialize to 100
        // baseline before halving so the player actually has HP.
        if (state.playerMaxHp <= 0) state.playerMaxHp = 100;
        state.playerMaxHp = Math.floor(state.playerMaxHp / 2);
        state.playerHp = Math.min(state.playerHp, state.playerMaxHp);
        break;
      case "double_alarm":
        state.alarmSpeedMultiplier = 2.0;
        break;
      case "ice_x2":
        state.encounterMultiplier = 2;
        break;
      case "no_heal":
        state.healDisabled = true;
        break;
      case "stealth_only":
        state.skillFilter = "stealth_only";
        break;
    }
  }
  state.activeMutators = Object.freeze(mutators.slice());
}

/** Clear all mutator effects from a run state. */
export function clearMutators(state: MutableRunState): void {
  for (const mutator of state.activeMutators) {
    switch (mutator) {
      case "low_hp":
        // Only restore if currently halved (not zero).
        if (state.playerMaxHp > 0) state.playerMaxHp = state.playerMaxHp * 2;
        break;
      case "double_alarm":
        state.alarmSpeedMultiplier = 1.0;
        break;
      case "ice_x2":
        state.encounterMultiplier = 1;
        break;
      case "no_heal":
        state.healDisabled = false;
        break;
      case "stealth_only":
        state.skillFilter = null;
        break;
    }
  }
  state.activeMutators = Object.freeze([]);
}

/** True if the given mutator is currently active. */
export function isMutatorActive(state: MutableRunState, mutator: RunMutator): boolean {
  return state.activeMutators.includes(mutator);
}

/** Return the active mutators list (defensive copy). */
export function getActiveMutators(
  state: MutableRunState,
): ReadonlyArray<RunMutator> {
  return state.activeMutators.slice();
}

/** Current alarm speed multiplier (1.0 = normal, 2.0 = DOUBLE_ALARM). */
export function getAlarmMultiplier(state: MutableRunState): number {
  return state.alarmSpeedMultiplier;
}

/** Current encounter count multiplier (1 = normal, 2 = ICE_X2). */
export function getEncounterMultiplier(state: MutableRunState): number {
  return state.encounterMultiplier;
}

/** True if HEAL salvage is disabled (NO_HEAL mutator). */
export function isHealDisabled(state: MutableRunState): boolean {
  return state.healDisabled;
}

/** True if only stealth skills are available (STEALTH_ONLY mutator). */
export function isStealthOnly(state: MutableRunState): boolean {
  return state.skillFilter === "stealth_only";
}

/**
 * Construct a fresh, default {@link MutableRunState}.
 *
 * Mirrors the Python baseline (`player_max_hp = 100`, alarm ×1, encounter
 * ×1, heal enabled, no skill filter, no active mutators).
 */
export function makeDefaultMutableRunState(): MutableRunState {
  return {
    playerHp: 100,
    playerMaxHp: 100,
    alarmSpeedMultiplier: 1.0,
    encounterMultiplier: 1,
    healDisabled: false,
    skillFilter: null,
    activeMutators: Object.freeze([]),
  };
}
